llama-index-integrations/vector_stores/llama-index-vector-stores-mariadb/llama_index/vector_stores/mariadb/base.py

```diff
@@ -0,0 +1,445 @@
+            connection_args=conn_args,
+            table_name=table_name,
+            schema_name=schema_name,
+            embed_dim=embed_dim,
+            perform_setup=perform_setup,
+            debug=debug,
+        )
+
+    @property
+    def client(self) -> Any:
+        if not self._is_initialized:
+            return None
+        return self._engine
+
+    def _connect(self) -> Any:
+        self._engine = sqlalchemy.create_engine(
+            self.connection_string, connect_args=self.connection_args, echo=self.debug
+        )
+
+    def _create_table_if_not_exists(self) -> None:
+        with self._engine.connect() as connection:
+            stmt = f"""
+            CREATE TABLE IF NOT EXISTS {self.table_name} (
+                id SERIAL PRIMARY KEY,
+                node_id VARCHAR(255) NOT NULL,
+                text TEXT,
+                metadata JSON,
+                embedding BLOB NOT NULL,
+                VECTOR INDEX (embedding)
+            );
+            """
+            connection.execute(sqlalchemy.text(stmt))
+
+            connection.commit()
+
+    def _initialize(self) -> None:
+        if not self._is_initialized:
+            self._connect()
+            if self.perform_setup:
+                self._create_table_if_not_exists()
+            self._is_initialized = True
+
+    def get_nodes(
+        self,
+        node_ids: Optional[List[str]] = None,
+        filters: Optional[MetadataFilters] = None,
+    ) -> List[BaseNode]:
+        """Get nodes from vector store."""
+        self._initialize()
+
+        stmt = f"""SELECT text, metadata FROM {self.table_name} WHERE node_id IN :node_ids"""
+
+        with self._engine.connect() as connection:
+            result = connection.execute(sqlalchemy.text(stmt), {"node_ids": node_ids})
+
+        nodes: List[BaseNode] = []
+        for item in result:
+            node = metadata_dict_to_node(json.loads(item.metadata))
+            node.set_content(str(item.text))
+            nodes.append(node)
+
+        return nodes
+
+    def _node_to_table_row(self, node: BaseNode) -> Dict[str, Any]:
+        return {
+            "node_id": node.node_id,
+            "text": node.get_content(metadata_mode=MetadataMode.NONE),
+            "embedding": node.get_embedding(),
+            "metadata": node_to_metadata_dict(
+                node,
+                remove_text=True,
+                flat_metadata=self.flat_metadata,
+            ),
+        }
+
+    def add(
+        self,
+        nodes: List[BaseNode],
+        **add_kwargs: Any,
+    ) -> List[str]:
+        self._initialize()
+
+        ids = []
+        with self._engine.connect() as connection:
+            for node in nodes:
+                ids.append(node.node_id)
+                item = self._node_to_table_row(node)
+                stmt = sqlalchemy.text(
+                    f"""
+                INSERT INTO {self.table_name} (node_id, text, embedding, metadata)
```

Comment:
Can we make `self.table_name` a sql variable instead of directly including in the SQL string?

Reply:
I tried to do it and the queries failed.
After quick googling it seems that MariaDB (and MySQL) doesn't support table names as variables (dynamic SQL).